src/Fusonic/Linq/Iterator/ChunkIterator.php

```diff
@@ -0,0 +1,98 @@
+<?php
+
+/*
+ * This file is part of Fusonic-linq.
+ * https://github.com/fusonic/fusonic-linq
+ *
+ * (c) Fusonic GmbH
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Fusonic\Linq\Iterator;
+
+use Fusonic\Linq\Linq;
+use Iterator;
+
+/**
+ * Iterates over an iterator, returning Linq objects of the given chunk size.
+ */
+class ChunkIterator implements Iterator
+{
+    /**
+     * @var Iterator
+     */
+    private $iterator;
+
+    /**
+     * @var array
+     */
+    private $chunk;
+
+    /**
+     * @var int
+     */
+    private $i = 0;
+
+    /**
+     * @var int
+     */
+    private $chunkSize;
+
+    public function __construct(Iterator $iterator, $chunkSize)
+    {
+        $this->iterator = $iterator;
+        $this->chunkSize = $chunkSize;
+
+        $this->chunk = $this->getNextChunk();
```

Comment:
Cool, well I submitted an update. I think this is why I always got slightly lost trying to implement iterators in PHP previously. Lots of mutable state and not tooo much indication of correct usage. Probably a better design would make incorrect usage syntactically impossible, but I doubt PHP will ever get that mathematical :)